npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb

```diff
@@ -2,18 +2,213 @@
+
+    PACKAGE_MANAGER_CLASSES = {
+      NpmPackageManager::NAME => NpmPackageManager,
+      YarnPackageManager::NAME => YarnPackageManager,
+      PNPMPackageManager::NAME => PNPMPackageManager
+    }.freeze
+
+    class PackageManagerDetector
+      extend T::Sig
+      extend T::Helpers
+
+      sig do
+        params(
+          lockfiles: T::Hash[Symbol, T.nilable(Dependabot::DependencyFile)],
+          package_json: T::Hash[String, T.untyped]
+        ).void
+      end
+      def initialize(lockfiles, package_json)
+        @lockfiles = lockfiles
+        @package_json = package_json
+        @manifest_package_manager = package_json["packageManager"]
+        @engines = package_json.fetch("engines", nil)
+      end
+
+      sig { returns(T.nilable(String)) }
+      def detect_package_manager
+        name_from_lockfiles || name_from_package_manager_attr || name_from_engines
+      end
+
+      private
+
+      sig { returns(T.nilable(String)) }
+      def name_from_lockfiles
+        # We prioritize the package manager that has a lockfile
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @lockfiles[manager_name.to_sym]
+        end
+        nil
+      end
+
+      sig { returns(T.nilable(String)) }
+      def name_from_package_manager_attr
+        return unless @manifest_package_manager
+
+        # We prioritize the package manager that has a lockfile
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @manifest_package_manager.start_with?("#{manager_name}@")
+        end
+      end
+
+      sig { returns(T.nilable(String)) }
+      def name_from_engines
+        return unless @engines.is_a?(Hash)
+
+        PACKAGE_MANAGER_CLASSES.each_key do |manager_name| # iterates keys in order as defined in the hash
+          return manager_name.to_s if @engines[manager_name.to_s]
+        end
+      end
+    end
+
+    class PackageManagerHelper
       extend T::Sig
       extend T::Helpers
+
+      DEFAULT_PACKAGE_MANAGER = NpmPackageManager::NAME
+
+      sig do
+        params(
+          package_json: T::Hash[String, T.untyped],
+          lockfiles: T::Hash[Symbol, T.nilable(Dependabot::DependencyFile)]
+        ).void
+      end
       def initialize(package_json, lockfiles:)
         @package_json = package_json
         @lockfiles = lockfiles
-        @package_manager = package_json.fetch("packageManager", nil)
+        @manifest_package_manager = package_json["packageManager"]
         @engines = package_json.fetch("engines", nil)
+        @package_manager_detector = PackageManagerDetector.new(@lockfiles, @package_json)
+      end
+
+      sig { returns(Ecosystem::VersionManager) }
+      def package_manager
+        name = @package_manager_detector.detect_package_manager || DEFAULT_PACKAGE_MANAGER
+        package_manager_by_name(name)
       end
```

Comment:
After detection package manager we want to return the created package manager information that can be used for metrics collection or other purposes.